### DISCUSSION TITLE
Public API

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#3542] [Ideas]
There have been a few discussions/issues about documentation for available commands.

A lot of this confusion arises because the `lazy` function calls are different to the available methods in the `qtile` object. Some of that confusion will be fixed by #2727 but there is a broader question as to whether or not users should be accessing the `qtile` object directly.

To be clear: I'm not telling users to stop doing this. It's python so it's very hackable and users should be able to do what they want but we need the flexibility to change underlying code from time to time.

My idea was that we could expose a `CommandClient` for the qtile objects which would allow users to carry on using the same command formats that they have been using for `lazy` commands. That familiarity could be beneficial.

The way it works is as follows: every `CommandObject` (`qtile`, windows, widgets, groups, layouts, screens, backends) expose a `c` attribute which is an instance of an `InteractiveCommandClient` based on the current object.

Examples:

``` python
@lazy.function
def my_lazy_function(qtile):
    qtile.c.reload_config()   # same as lazy.reload_config()

@lazy.widget["textbox"].function
def update_textbox_text(widget):
    new_text = "New message"
    widget.c.update(new_text)   # same as lazy.widget["textbox"].update("New message")

# Functions can then be attached to keybindings
```

We only need about 10 lines of code to add this but I'm curious if people think it would be useful or not.

**Comment by ervinpopescu** [comment]
@elParaguayo this would be nice to have, just a reminder :)

**Comment by bodograumann** [comment]
To be honest, for me as a new user the different ways to interact with qtile in the configuration through the predefined lazy commands, defining lazy function commands, using hooks and doing things via qtile shell are extremely confusing. Sometimes I have to prefix with cmd_ and what is actually available is pretty unclear. Maybe there are even more ways to interact with qtile than I am aware of.
So any way to smooth out these differences would be highly appreciated.
Of course the documentation could be improved arround this, but I'd prefer not having to learn about such differences in the first place.

**Reply by m-col** [reply]
Just replying to acknowledge and say thanks for the feedback.

With https://github.com/qtile/qtile/pull/2727 merged, the next release will see the `cmd_` prefix dropped completely and therefore any `lazy.<command>` will directly translate to `qtile.<command>`. There are more changes that could further align these two interfaces (e.g. changing `qtile.current_window` to `qtile.window`, which would make `lazy.window.<command>` map directly to `qtile.window.<command>`).

To respond to the initial post (apologies, @elParaguayo discussed this elsewhere but didn't update this discussion), we decided against the `CommandObject` approach described above for 2 reasons:

1. unnecessary runtime work to execute the commands
2. The 1:1 mapping of `lazy.X` to `qtile.X` with an arbitrary path (`qtile.window.screen.bar` etc) I just mentioned can be accomplished which (IMO) would be more straight forward than having to put `.c` within it (`qtile.c.window.screen.bar`).

That would address the stated problem:

> A lot of this confusion arises because the lazy function calls are different to the available methods in the qtile object.

Then there is your description:

> Maybe there are even more ways to interact with qtile than I am aware of.
> So any way to smooth out these differences would be highly appreciated.
> Of course the documentation could be improved arround this, but I'd prefer not having to learn about such differences in the first place.

Having the 1:1 mapping eliminates the differences, but only for _commands_. The remaining question is how to expose and document any other attributes that people sometimes access on the `qtile` object (or other objects for that matter).

**Reply by elParaguayo** [reply]
Thanks m-col.

I think exposing other objects/attributes would be relatively easy (we could easily add another decorator to mark those items that we want to have picked up in our docs). However, before we do that, we need to agree on what we expose because, once we've done that, it's effectively public API and harder for us to change.

I do like the idea of aligning the `lazy.` and `qtile.` syntax as the `current_` part can be confusing.

Another thought I had was whether we could just add a `.call()` method to lazy objects which would make them executable in their own right and so the same syntax could be used within functions e.g.
```python
def my_function():
    lazy.spawn("firefox").call()
```
It's probably unnecessary but I'm just trying to think of ways we can minimise the confusion. The problem with the above approach is that there's now confusion between the lazy syntax used in keybindings and the syntax used in functions!
